src/server.c

```diff
@@ -346,13 +350,25 @@ static bool handle_server_work(PgSocket *server, PktHdr *pkt)
 		break;
 	/* chat packets */
 	case '2':		/* BindComplete */
+		break;
 	case '3':		/* CloseComplete */
+		if (is_prepared_statements_enabled(pool))
+			ignore_packet = true;
```

Comment:
Yes, in general you can't just ignore close packets, hence the original `FIXME`. As our specific setup does not use unnamed prepared statements or portals we never had the need to fix this properly...

Just happens I gave this issue some thought earlier today. My idea atm is to keep a stack of every Close packet sent to the server, and popping the stack on receiving a CloseComplete packet. Only forwarding the CloseComplete to the client if it is not in response to the Close packets we generate to cleanup the server connection. Doing so ensures the order of packets received by the client, which is hard to do with just a counter I think.

Reply:
Yes, you're right a counter cannot hold the order indeed. So we need a stack similar to the `server_outstanding_parse_packets` we have for Parse packets currently.

Reply:
Actually even then you can get reordering of results too. By using `pktbuf_send_immediate` we return CloseComplete and ParseComplete before possible other packets that the client still expected from the server. I'm not sure if there's an (easy) way to fix that though.

Reply:
I added `skip_close_response` to `PgSocket` and turn it  on in  `register_prepared_statement` when CLOSE request is sent to the server. Only in this case '3' is ignored.
Should it work?

Reply:
So a boolean would be fine if there was only ever one request in flight. But that's not necessarily the case in case of request pipelining: https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-PIPELINING

This PR introduces multiple issues for pipelining though. And I don't think it's trivial to fix all of them. I think for the initial version of this prepared statement features it would be best to document that we don't support pipelining, and not make it too complicated. One thing that does seem rather simple is to use a counter instead of a boolean, and finally an important thing is to reset this counter to 0 when we receive ErrorResponse from the server. Because that means that postgres won't respond to any of our in flight parse messages, because the transaction was cancelled. 

The stack proposed by @sboschman would indeed solve a few more issues, but I don't think it's worth it in the complexity, definitely not for these Close packets at least. @sboschman @dashorst did you use the stack for outstanding parse packets.


Actually fixing pipelining + prepared statements would require us to insert all the packets that we currently send with `pktbuf_send_immediate` to instead insert into the stream of packets from postgres at the right place. The only way in which we can do this is probably by keeping track of all in flight packets (not just Parse and Close) and matching those with the responses from Postgres to pop them appropriately from the stack. Then instead of using pktbuf_send_immediate, we would add "fake" items to the in flight request stack. And everytime we pop from the stack because of a response from Postgres, we also peak to the next entry to see if that's a "fake" item and if it is we pop it and we send that response right away. But like I said I don't think this is necessary to get this PR merged. This PR would still fix prepared statements for everyone that does not use pipelining. And even with pipelining, most of the time everything should work correctly too.

Reply:
I updated the PR to now maintain a stack, of the packets that libpq cares about receiving in order when using pipelining.